Mathlib/Algebra/GCDMonoid/Multiset.lean

```diff
@@ -0,0 +1,279 @@
+Authors: Aaron Anderson
+
+! This file was ported from Lean 3 source module algebra.gcd_monoid.multiset
+! leanprover-community/mathlib commit 9003f28797c0664a49e4179487267c494477d853
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Algebra.GCDMonoid.Basic
+import Mathlib.Data.Multiset.FinsetOps
+import Mathlib.Data.Multiset.Fold
+
+/-!
+# GCD and LCM operations on multisets
+
+## Main definitions
+
+- `Multiset.gcd` - the greatest common denominator of a `Multiset` of elements of a `GCDMonoid`
+- `Multiset.lcm` - the least common multiple of a `Multiset` of elements of a `GCDMonoid`
+
+## Implementation notes
+
+TODO: simplify with a tactic and `Data.Multiset.Lattice`
+
+## Tags
+
+multiset, gcd
+-/
+
+namespace Multiset
+
+variable {α : Type _} [CancelCommMonoidWithZero α] [NormalizedGCDMonoid α]
+
+/-! ### lcm -/
+
+
+section Lcm
+
+/-- Least common multiple of a multiset -/
+def lcm (s : Multiset α) : α :=
+  s.fold GCDMonoid.lcm 1
+#align multiset.lcm Multiset.lcm
+
+@[simp]
+theorem lcm_zero : (0 : Multiset α).lcm = 1 :=
+  fold_zero _ _
+#align multiset.lcm_zero Multiset.lcm_zero
+
+@[simp]
+theorem lcm_cons (a : α) (s : Multiset α) : (a ::ₘ s).lcm = GCDMonoid.lcm a s.lcm :=
+  fold_cons_left _ _ _ _
+#align multiset.lcm_cons Multiset.lcm_cons
+
+@[simp]
+theorem lcm_singleton {a : α} : ({a} : Multiset α).lcm = normalize a :=
+  (fold_singleton _ _ _).trans <| lcm_one_right _
+#align multiset.lcm_singleton Multiset.lcm_singleton
+
+@[simp]
+theorem lcm_add (s₁ s₂ : Multiset α) : (s₁ + s₂).lcm = GCDMonoid.lcm s₁.lcm s₂.lcm :=
+  Eq.trans (by simp [lcm]) (fold_add _ _ _ _ _)
+#align multiset.lcm_add Multiset.lcm_add
+
+theorem lcm_dvd {s : Multiset α} {a : α} : s.lcm ∣ a ↔ ∀ b ∈ s, b ∣ a :=
+  Multiset.induction_on s (by simp)
+    (by simp (config := { contextual := true }) [or_imp, forall_and, lcm_dvd_iff])
+#align multiset.lcm_dvd Multiset.lcm_dvd
+
+theorem dvd_lcm {s : Multiset α} {a : α} (h : a ∈ s) : a ∣ s.lcm :=
+  lcm_dvd.1 dvd_rfl _ h
+#align multiset.dvd_lcm Multiset.dvd_lcm
+
+theorem lcm_mono {s₁ s₂ : Multiset α} (h : s₁ ⊆ s₂) : s₁.lcm ∣ s₂.lcm :=
+  lcm_dvd.2 fun _ hb ↦ dvd_lcm (h hb)
+#align multiset.lcm_mono Multiset.lcm_mono
+
+/- Porting note: Following `Algebra.GCDMonoid.Basic`'s version of `normalize_gcd`, I'm giving
+this lower priority to avoid linter complaints about simp-normal form -/
+/- Porting note: Mathport seems to be replacing `multiset.induction_on s $` with
+`(Multiset.induction_on s)`, when it should be `Multiset.induction_on s <|`. -/
+@[simp 1100]
+theorem normalize_lcm (s : Multiset α) : normalize s.lcm = s.lcm :=
+  Multiset.induction_on s (by simp) <| fun a s _ ↦ by simp
+#align multiset.normalize_lcm Multiset.normalize_lcm
+
+/- Porting note: When trying to use `simp only [lcm_eq_zero_iff]`, Lean seems to think
+`lcm_eq_zero_iff` refers to the very theorem we are trying to prove. -/
```

Comment:
```suggestion
```
This is intentional. So it doesn't need a porting note. Either fix it with `nonrec`, as you did, or by prefixing `lcm_eq_zero_iff` in the proof with a namespace.